### ISSUE TITLE
pq.write_to_dataset possible memory leak

### ISSUE THREAD

**Issue by asafcombo** [apache/arrow#1285]
# Description

running `write_to_dataset `  seems to keep a high memory footprint after function completion. 
   This hinders running subsequent calls to the function.

# Replication
1 . Running on aws `t2.large`.
2. 
```python
import random
import pandas as pd
import pyarrow as pa
import pyarrow.parquet as pq

array_to_pd = []
temp_dic['a'] = "z"*9
temp_dic['b'] = "y"*9
temp_dic['c'] = "x"*9
for i in range(2000000):
    temp_dic['d'] = random.randint(1,20)
    temp_dic['e'] = random.randint(1,100)
    temp_dic['url'] = 'www.google.com' + str(random.randint(1,30000))
    array_to_pd += [temp_dic.copy()]

df = pd.DataFrame.from_dict(array_to_pd,orient='columns')
table = pa.Table.from_pandas(df)
pq.write_to_dataset(table,'/demo/',['d','e'])

# memory utilization stays high - running again will crush program
pq.write_to_dataset(table,'/demo/',['d','e'])
```

**Comment by xhochy** [comment]
Yes, I can reproduce this problem locally with the current git master. This is not a case of a real memory leak but a cyclic reference somewhere in our implementation. Explicitly calling `gc.collect()` after `from_pandas` and `write_to_dataset` keeps the memory usage at a low level. While this is a workaround that should get you ahead with your code, in the long term we should remove the cyclic references so that explicit GC calls are not necessary anymore.

**Comment by xhochy** [comment]
I opened https://issues.apache.org/jira/browse/ARROW-1769 about this issue.

**Comment by asafcombo** [comment]
@xhochy  ok, will try the workaround.
Appreciate the help with handling both questions

**Comment by xhochy** [comment]
@asafcombo I will close this here as there seems to be nothing we can do immediately to fix this inside of `pyarrow`. Updating to the current git-master of pandas fixes the problem (this will be released as Pandas 0.22 in the future, probably around January). In the JIRA issues, we might discuss to build a workaround but adding a `gc.collect()` after `write_dataset` is not the cleanest but most efficient for the moment.
